device/src/async_device/mod.rs

```diff
@@ -221,6 +221,13 @@ where
             .region
             .create_tx_config(&mut self.rng, self.datarate, &Frame::Data);
 
+        // Unless the same frame is to be retransmitted (see NbTrans parameter of LinkADRReq command, LoRaWAN spec
+        // 1.0.2 section 5.2 for retransmissions), FCnt must be incremented on each transmission.
+        self.session
```

Comment:
FWIW, in the sync stack, we wait increment FCnt in the "WaitingForRx" state:

- If a packet is received, we always increment the FCntUp; if this was a confirmed or unconfirmed frame, no matter.
- If the Rx2 window expires and it was an unconfirmed uplink, then we also increment.

So the issue with this implementation is that if it's a confirmed uplink with no confirmation, you've still incremented the FCntUp. This might make it difficult to implement retry on lack of confirmation, for example.